modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationParameterValueResolver.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.operation;
+
+import org.mule.runtime.api.meta.model.operation.OperationModel;
+import org.mule.runtime.extension.api.runtime.operation.OperationContext;
+import org.mule.runtime.module.extension.internal.introspection.ParameterGroup;
+import org.mule.runtime.module.extension.internal.model.property.ParameterGroupModelProperty;
+import org.mule.runtime.module.extension.internal.runtime.ParameterValueResolver;
+import org.mule.runtime.module.extension.internal.runtime.resolver.ParameterGroupArgumentResolver;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * {@link ParameterValueResolver} implementation for {@link OperationMessageProcessor}
+ *
+ * @since 4.0
+ */
+public class OperationParameterValueResolver implements ParameterValueResolver {
```

Comment:
this class doesn't make sense:

* First it's not for OperationMessageProcessor, since it's only value seems to be the `resolveSingleLevelMetadataKey` method which is Metadata specific
* Secondly, it takes an OperationContext, which means that this is more a wrapper than anything else
* It contains parameter group logic, which is also a smell

In summary, this looks like a big hack